### ISSUE TITLE
DNA bundle format and lineage definition

### ISSUE THREAD

**Issue by abe-njama** [holochain/holochain#3955]
Make it possible to depend on a DNA from another app at install time:

- New field in the _DNA manifest_ to specify a list of DNA hashes (= lineage) that are compatible (coordinator interface) with the current one. Guaranteeing lies with the happ developer. Holochain does not check or enforce compatibility in any way.
- There's a function that takes a single DNA hash and returns the latest DNA in that lineage if the lineage is registered.

The DNA lineage is a vector of DNA hashes, ordered by version in descending order (latest version first). When installing an app and `use_existing` is encountered as a provisioning strategy of a DNA, the conductor will iterate over the vector of DNA hashes and search existing DNAs.

Tests need to ensure the lookup of the right existing DNA and test cases where it is found and where it's not found.

**Comment by jost-s** [comment]
- Why a lineage of DNA hashes? Is this like a set?
- In what way is `hc_bundle` based on the previous happ?
- Which other `use_existing` variants could/should there be?

**Comment by jost-s** [comment]
- Forward/backward compatibility required?
- Coordinator interface compatibility guarantees?
- DNA hash authenticity check?

We need to clarify what are the immediate requirements.

**Comment by jost-s** [comment]
Each DNA specifies its compatibility with different versions of itself. Any app just specifies a hash of a targeted version of a DNA. If that hash is found in any lineage of the existing DNAs (all installed DNAs), the DNA of this matching lineage is taken for app installation. For this devs need to know the DNA hash of some version of the DNA they're including in their app manifest.

In the case of multiple matching lineages for a DNA in the conductor, the conductor can't make a decision. The user needs to pick one or abandon the install. Return a structured error, describing all the matches (so not just latest DNA hash for each option, but also installed app id and agent key). Let the caller retry the installation with some way to specify which option to choose for each multi-choice dep.

Example:
Say I am building an app, and I have produced versions with DNA hashes H1, H2, H3, H4, H5.

I have decided that H3 was a bad version and I want people to not use it. But otherwise I have written migrations in each version so that H2 knows how to migrate from H1. The pattern continues up to H5.

I have kept my data structures and coordinators 'compatible' according to some definition that I'm sharing with people building on top of my app.

So, when I go to release H5, I want to describe the lineage of my DNA. In my DNA manifest I specify [H1, H2, H4] as the lineage. I don't need to specify myself because I am H5. Ideally I'd have some CLI tooling to help me with this packaging step, but that's something for later!

Now when people install my app into a conductor, the lineage stays with the app because it was part of the DNA definition.

Later on, the person who installed my app tries to use an app built on top of mine that was originally compatible with H2. In the DNA manifest for this other app, the developer has specified a use_existing which references H2. On installation, the DNA H2 would not be found, because H5 is running. However, the conductor can search the lineage for my app and discover the sequence from H2. This means the conductor can understand that there's a possibility of compatibility and proceed with the install against my app.

The other app may specify its own lineage for a further depth of dependent installs. But in order to link with my app, it only specifies a single hash. 

The point of what we're trying to achieve with this is expanding the range of apps that can be installed together without prior knowledge of exact versions. If we did everything by hash we'd end up in a horrible situation where apps are in a constant state of repackaging every time small features are added. By making the conductor more flexible, we allow users to make their own decisions about compatibility. If we required this other app to specify a range of versions then that would defeat the purpose of adding flexibility to some extent.
